# 前言
这是一个跨国数据集，其中包含在 2010 年 1 月 12 日到 2011 年 9 月 12 日之间发生的英国某电商在线零售的交易数据。
数据量很庞大，在分析思路上可以使用机器学习K-Means 等算法，根据客户在市场上的购买行为来细分客户。

Syntax	Description
InvoiceNo	发票号码，一个唯一分配给每笔交易的 6 位整数。如果此代码以字母“c”开头，则表示取消
StockCode	股票代码
Description	描述，产品（项目）名称
Quantity	数量
InvoiceDate	发票日期
UnitPrice	单价，以英镑为单位的每单位产品价格
CustomerID	客户ID
Country	国家
————————————————

                            
                        
#原文链接：https://blog.csdn.net/weixin_45058606/article/details/126069838

## 读取数据

导入数据分析需要的库，并通过Pandas的read_scv函数，将原始数据"Online Retail.csv"里的数据内容，解析为DataFrame,并赋值给original_data

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("Online Retail.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  2010/12/1 8:26       2.55     17850.0  United Kingdom  
1  2010/12/1 8:26       3.39     17850.0  United Kingdom  
2  2010/12/1 8:26       2.75     17850.0  United Kingdom  
3  2010/12/1 8:26       3.39     17850.0  United Kingdom  
4  2010/12/1 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的 original_data 这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据整齐度

In [4]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
508005   C579192     22809             SET OF 6 T-LIGHTS SANTA       -12   
398135    571221     21385     IVORY HANGING DECORATION  HEART         2   
302317    563425     23296      SET OF 6 TEA TIME BAKING CASES        16   
23847     538255     21042                RED RETROSPOT APRON          3   
188963    553074     22998        TRAVEL CARD WALLET KEEP CALM         4   
193280    553523     21668      RED STRIPE CERAMIC DRAWER KNOB         4   
192882    553487     22219  LOVEBIRD HANGING DECORATION WHITE          7   
106846    545346     22283            6 EGG HOUSE PAINTED WOOD         2   
73033     542277    84970L    SINGLE HEART ZINC T-LIGHT HOLDER        12   
353124    567742     23107  WHITE HEARTS WIRE PLANT POT HOLDER         8   

             InvoiceDate  UnitPrice  CustomerID         Country  
508005  2011/11/28 15:43       2.95     12657.0          France  
398135  2011/10/14 13:35       0.85     16416.0  United Kingdom  
302317   2011/8/16 11:37       1.25     15140.0  United Kingdom  
23847   2010/12/10 12:33       5.95     14911.0            EIRE  
188963   2011/5/13 11:45       0.42     15529.0  United Kingdom  
193280   2011/5/17 14:46       2.46         NaN  United Kingdom  
192882   2011/5/17 12:38       0.85     13246.0  United Kingdom  
106846     2011/3/2 9:49       7.95     12834.0  United Kingdom  
73033    2011/1/27 10:30       0.95     14440.0  United Kingdom  
353124   2011/9/22 10:47       2.89     14261.0  United Kingdom


从抽样结果来看，符合每行是一个变量，每列是一个观察者，每个单元格是一个值。具体来看每行是商品的一次交易，每列是相关的变量，因此不存在结构性问题

### 评估数据的干净程度

In [5]:
 original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541,909条观察值，而Description、CustomerID变量存在缺失值。

此外，InvoiceDate的数据类型应为日期，CustomerID的数据类型应为字符串，应当进行数据格式转换。

### 评估缺失数据

在了解到description存在缺失值后，根据条件提取带有缺失值的行

In [6]:
original_data[original_data['Description'].isna()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  2010/12/1 11:52        0.0   
1970      536545     21134         NaN         1  2010/12/1 14:32        0.0   
1971      536546     22145         NaN         1  2010/12/1 14:33        0.0   
1972      536547     37509         NaN         1  2010/12/1 14:33        0.0   
1987      536549    85226A         NaN         1  2010/12/1 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  2011/12/7 18:26        0.0   
535326    581203     23406         NaN        15  2011/12/7 18:31        0.0   
535332    581209     21620         NaN         6  2011/12/7 18:35        0.0   
536981    581234     72817         NaN        27  2011/12/8 10:33        0.0   
538554    581408     85175         NaN        20  2011/12/8 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失Description变量值。

从输出结果来看，这些缺失Description的交易数据，UnitPrice都为0。为了验证猜想，我们增加筛选条件，看是否存在Description变量缺失且UnitPrice不为0的数据。

In [7]:
original_data[(original_data['Description'].isna())&(original_data['UnitPrice'] != 0.0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选结果显示0条记录，说明缺失Description值的数据同时也不具备有效的UnitPrice值。

Description（产品名称）和UnitPrice（产品单价）是商品交易分析的关键变量。当这两个字段同时缺失/无效时，数据无法提供有效信息，建议删除这些记录。

-----------------
CustomerID同意存在缺失值

In [8]:
original_data[original_data['CustomerID'].isna()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     2010/12/1 11:52       0.00         NaN  United Kingdom  
1443    2010/12/1 14:32       2.51         NaN  United Kingdom  
1444    2010/12/1 14:32       2.51         NaN  United Kingdom  
1445    2010/12/1 14:32       0.85         NaN  United Kingdom  
1446    2010/12/1 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  2011/12/9 10:26       4.13         NaN  United Kingdom  
541537  2011/12/9 10:26       4.13         NaN  United Kingdom  
541538  2011/12/9 10:26       4.96         NaN  United Kingdom  
541539  2011/12/9 10:26      10.79         NaN  United Kingdom  
541540  2011/12/9 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID（客户编号）不是分析畅销商品的必要变量。根据输出结果显示，部分CustomerID缺失的交易数据仍然有效，因此保留这些CustomerID为空的观察值。

### 评估重复数据

InvoiceNo（发票编号）：单次交易可能包含多件商品，因此允许重复

StockCode（商品编码）：不同交易可能包含相同商品，因此允许重复

CustomerID（客户编号）：同一客户可进行多次交易，因此允许重复

### 评估不一致数据

不一致数据可能存在Country变量中，我们要查看是否存在多个值指代同一个国家的情况

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


美国存在多种表示："USA"、"United States"

英国存在多种表示："United Kingdom"、"UK"、"U.K."

### 评估无效或错误数据

可以通过DataFrame的`describe`方法，对数值的统计信息进行快速的了解

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

退货订单通常表现为负Quantity

价格补偿/调整可能产生负UnitPrice

In [11]:
original_data[original_data['Quantity'] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      2010/12/1 9:41      27.50     14527.0  United Kingdom  
154      2010/12/1 9:49       4.65     15311.0  United Kingdom  
235     2010/12/1 10:24       1.65     17548.0  United Kingdom  
236     2010/12/1 10:24       0.29     17548.0  United Kingdom  
237     2010/12/1 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   2011/12/9 9:57       0.83     14397.0  United Kingdom  
541541  2011/12/9 10:28     224.69     15498.0  United Kingdom  
541715  2011/12/9 11:57      10.95     15311.0  United Kingdom  
541716  2011/12/9 11:58       1.25     17315.0  United Kingdom  
541717  2011/12/9 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，Quantity变量为负数的观察值，InvoiceNo似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

In [12]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  2010/12/1 16:50        0.0   
4347      536764    84952C         NaN       -38  2010/12/2 14:42        0.0   
7188      536996     22712         NaN       -20  2010/12/3 15:30        0.0   
7189      536997     22028         NaN       -20  2010/12/3 15:30        0.0   
7190      536998     85067         NaN        -6  2010/12/3 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  2011/12/7 18:36        0.0   
535335    581212     22578        lost     -1050  2011/12/7 18:38        0.0   
535336    581213     22576       check       -30  2011/12/7 18:38        0.0   
536908    581226     23090     missing      -338   2011/12/8 9:56        0.0   
538919    581422     23169     smashed      -235  2011/12/8 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但根据输出结果，这些筛选出的观察值的UnitPrice观察值均为0，因此需要增加UnitPrice的条件进行进一步验证。

In [13]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != "C") & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据验证结果，Quantity为负数的记录符合以下特征之一：

InvoiceNo以"C"开头（表示取消的订单）

UnitPrice为0（无效交易价格）

这些记录均不构成有效销售数据，将在清洗阶段被移除。

-----
接下来筛选UnitPrice也是负数的行

In [14]:
original_data[(original_data['UnitPrice'] < 0)]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  2011/8/12 14:51   
299984   A563187         B  Adjust bad debt         1  2011/8/12 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

根据分析结果，UnitPrice为负数的记录属于坏账调账（非实际商品交易数据），应在数据清洗中予以删除。

## 清洗数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：

- 把 InvoiceDate 变量的数据类型转换为为日期时间
- 把 CustomerID 变量的数据类型转换为字符串
- 把 Description 变量缺失的观察值删除
- 把 Country 变量值 "USA" 替换为 "United States"
- 把 Country 变量值 "UK" 、 "U.K." 替换为 "United Kingdom"
- 把 Quantity 变量值为负数的观察值删除
- 把 UnitPrice 变量值为负数的观察值删除

为了区分清理前后的数据，我们创建 cleaned_data 作为 original_data 的副本，后续所有清理操作将在 cleaned_data 上执行：

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  2010/12/1 8:26       2.55     17850.0  United Kingdom  
1  2010/12/1 8:26       3.39     17850.0  United Kingdom  
2  2010/12/1 8:26       2.75     17850.0  United Kingdom  
3  2010/12/1 8:26       3.39     17850.0  United Kingdom  
4  2010/12/1 8:26       3.39     17850.0  United Kingdom

把 InvoiceDate 变量的数据类型转换为为日期时间

In [16]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把 CustomerID 变量的数据类型转换为字符串

In [17]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

我们要把`CustomerID`结尾的`.0`删除

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把 Description 变量缺失的观察值删除

In [19]:
cleaned_data.dropna(subset=["Description"],inplace=True)
cleaned_data["Description"].isna().sum()

0

把 Country 变量值 "USA" 替换为 "United States"

In [20]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA":"United States"})

In [21]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把 Country 变量值 "UK" 、 "U.K." 替换为 "United Kingdom"

In [22]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom","U.K":"United Kingdom"})

In [23]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K"]))

0
0


把 Quantity 变量值为负数的观察值删除

In [24]:
cleaned_data = cleaned_data[(cleaned_data['Quantity'] >= 0)]

In [25]:
len(cleaned_data[(cleaned_data['Quantity'] <= 0)])

0

把 UnitPrice 变量值为负数的观察值删除

In [26]:
cleaned_data = cleaned_data[(cleaned_data['UnitPrice'] >= 0)]

In [27]:
len(cleaned_data[(cleaned_data['UnitPrice'] < 0)])

0

In [28]:
cleaned_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
312443    564353     23153     ZINC SWEETHEART WIRE LETTER RACK         4   
284290    561868     22558      CLOTHES PEGS RETROSPOT PACK 24          1   
37850     539484     84380    SET OF 3 BUTTERFLY COOKIE CUTTERS        24   
217376    555924     23159         SET OF 5 PANCAKE DAY MAGNETS         4   
482063    577405     22468         BABUSHKA LIGHTS STRING OF 10         4   
511383    579495     22577  WOODEN HEART CHRISTMAS SCANDINAVIAN        24   
476442    577018     22947           WOODEN ADVENT CALENDAR RED         6   
165030    550756     22617           BAKING SET SPACEBOY DESIGN         3   
370224    569132     22546                   MINI JIGSAW PURDEY         1   
294423    562715     23124   18PC WOODEN CUTLERY SET DISPOSABLE        25   

               InvoiceDate  UnitPrice CustomerID         Country  
312443 2011-08-24 15:08:00       3.75      15764  United Kingdom  
284290 2011-07-31 11:26:00       1.65      16813  United Kingdom  
37850  2010-12-19 15:44:00       1.25      15640  United Kingdom  
217376 2011-06-07 17:12:00       4.13          n  United Kingdom  
482063 2011-11-20 10:29:00       6.75      14116  United Kingdom  
511383 2011-11-29 15:34:00       0.29      13576  United Kingdom  
476442 2011-11-17 12:25:00       7.95      12971  United Kingdom  
165030 2011-04-20 12:02:00       4.95      14148  United Kingdom  
370224 2011-09-30 13:54:00       0.42      17757  United Kingdom  
294423 2011-08-08 15:34:00       2.46          n  United Kingdom

## 保存清理后的数据

In [29]:
cleaned_data.to_csv("Online Retail_cleaned.csv",index = False)

In [30]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [31]:
pd.read_csv("Online Retail_cleaned.csv").head()

C:\Users\lihon\AppData\Local\Temp\ipykernel_24880\3246538791.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("Online Retail_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom